src/MobileRedirectManager.php

```diff
@@ -0,0 +1,225 @@
+			'Opera Mini',
+			'Opera Mobi',
+		];
+
+		/**
+		 * Filters the list of user agents used to determine if the user agent from the current request is a mobile one.
+		 *
+		 * @since 1.6
+		 *
+		 * @param string[] $user_agents List of user agents.
+		 */
+		return apply_filters( 'amp_mobile_user_agents', $default_user_agents );
+	}
+
+	/**
+	 * Determine if mobile redirection is disabled for the browser session.
+	 *
+	 * @return bool True if disabled, false otherwise.
+	 */
+	public static function redirection_disabled() {
+		return isset( $_COOKIE[ self::DISABLED_COOKIE_NAME ] ) && '1' === $_COOKIE[ self::DISABLED_COOKIE_NAME ];
+	}
+
+	/**
+	 * Output the mobile redirection Javascript code.
+	 */
+	public static function add_mobile_redirect_script() {
+		?>
+		<script>
+			(function ( ampSlug, disabledCookieName, userAgents ) {
+				const currentUserAgent = navigator.userAgent;
+				const isMobile = userAgents.some( ( userAgent ) => currentUserAgent.includes( userAgent ) );
+
+				if ( isMobile ) {
+					document.addEventListener( 'DOMContentLoaded', () => {
+						// Show the mobile version switcher link once the DOM has loaded.
+						const siteVersionSwitcher = document.getElementById( 'site-version-switcher' );
+						if ( siteVersionSwitcher ) {
+							siteVersionSwitcher.hidden = false;
+						}
+
+						// Disable mobile redirection if the user opts for the mobile version.
+						const versionSwitchLink = document.getElementById( 'version-switch-link' );
+						if ( versionSwitchLink ) {
+							versionSwitchLink.addEventListener( 'click', () => {
+								document.cookie = `${disabledCookieName}=;path=/;samesite=strict${ 'https:' === location.protocol ? ';secure' : '' }`;
+							} )
+						}
+					} );
+				}
+
+				const mobileRedirectionDisabled = document.cookie
+						.split(';')
+						.some( ( item ) => `${ disabledCookieName }=1` === item.trim() );
+
+				// Short-circuit if mobile redirection is disabled.
+				if ( mobileRedirectionDisabled ) {
+					return;
+				}
+
+				const url = new URL( location.href );
+				if ( isMobile && ! url.searchParams.has( ampSlug ) ) {
+					window.stop(); // Stop loading the page! This should cancel all loading resources.
+
+					// Replace the current page with the AMP version.
+					url.searchParams.append( ampSlug, '1' );
+					location.replace( url.href );
+				}
+			} )(
+				<?php echo wp_json_encode( amp_get_slug() ); ?>,
+				<?php echo wp_json_encode( self::DISABLED_COOKIE_NAME ); ?>,
+				<?php echo wp_json_encode( self::get_user_agents() ); ?>
+			)
+		</script>
+		<?php
+	}
+
+	/**
+	 * Output the markup for the mobile version switcher.
+	 *
+	 * @param bool   $is_amp Modifies markup to be AMP compatible if true.
+	 * @param string $url    URL to canonical version of page.
+	 * @param string $text   Text for the anchor element.
+	 */
+	public static function add_mobile_version_switcher_markup( $is_amp, $url, $text ) {
+		$rel         = $is_amp ? 'amphtml' : 'noamphtml';
```

Comment:
To prevent mobile crawlers from getting stuck on the non-AMP version:

```suggestion
		$rel         = $is_amp ? 'amphtml' : 'nofollow noamphtml';
```